capstone-backend/src/main/java/com/google/sps/servlets/Utility.java

```diff
@@ -36,26 +44,185 @@ public static Firestore getFirestoreDb() throws IOException {
   }
 
-  public static JsonObject createRequestBodyJson(HttpServletRequest request) {
-    JsonObject jsonObject = new JsonObject();
-    StringBuffer jb = new StringBuffer();
-    String line = null;
+  public static JsonObject createRequestBodyJson(HttpServletRequest request) throws IOException {
+    String jsonObjectString = request.getReader().lines().collect(Collectors.joining());
+    JsonObject jsonObject = JsonParser.parseString(jsonObjectString).getAsJsonObject();
+    return jsonObject;
+  }
+
+  /**
+   * Query the given collection by id. 
+   * @param {collectionReference} reference to the appropriate database collection
+   * @param {request} request sent to the backend
+   * @param {response} response returned from the call
+   * @param {genericClass} a Generic class, used in casting documents from the database
+   * @return List<T> a singleton List of the Object that matches the ID in the request.
+   */
+  private static <T> List<T> getById(CollectionReference collectionReference, HttpServletRequest request, 
+      HttpServletResponse response, GenericClass<T> genericClass) throws IOException {
+    if (request.getParameterMap().size() > 1) {
+      System.err.println("Error: No other parameter can be sent with an ID");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    List<T> retrievedObjects = new ArrayList<>();
+    DocumentReference docRef = collectionReference.document(request.getParameter("id"));
+    ApiFuture<DocumentSnapshot> asyncDocument = docRef.get();
+    DocumentSnapshot document = null;
+    T item = null;
     try {
-      BufferedReader reader = new BufferedReader(new InputStreamReader((request.getInputStream())));
-      while ((line = reader.readLine()) != null) {
-       jb.append(line);
+      document = asyncDocument.get();
+      if (document.exists()) {
+        item = document.toObject(genericClass.getMyType());
+      } else {
+        System.err.println("Error: no such document!");
+        response.sendError(HttpServletResponse.SC_NOT_FOUND);
+        return null;
       }
-    } catch (Exception e) { 
+    } catch (Exception e) {
       System.err.println("Error: " + e);
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return null;
+    }
+    retrievedObjects.add(item);
+    return retrievedObjects;
+  }
+
+  /**
+   * Appends filters to the given query depending on fields in the request query 
+   * @param {genericClass} a Generic class, used in casting documents from the database
+   * @param {response} response returned from the call
+   * @param {it} iterator for the Map of query parameter names to values
+   * @param {query} query being modified, to eventually be used to query the database.
+   * @return Query the query being modified, to eventually be used to query the database.
+   */
+  private static <T> Query addToQuery(GenericClass<T> genericClass, HttpServletResponse response, 
```

Comment:
It looks like the logic in the `addToQuery` method here is quite similar to what's already in `getByField`. Is there an easy way to remove the duplicate code (e.g. I'm wondering if a for loop over the `request.getParameterMap()`as opposed to using an `Iterator` would help)?

Reply:
Similar to my answer above, creating the `Query` and adding to it are different, and there is no way to create a base, empty Query to build off of. This forced me to duplicate some code, but I couldn't come up with a nice way around it without running into issues. 

For example, if I used a `for` loop and gave it different behavior on the first loop (instantiate Query on `i == 0`) than on the rest (add to Query on `i != 0`) I received errors along the lines of "Query may not have been instantiated" because there is no way to initialize the Query before the loop. 

I do think that the code here appears messy, and I'm open to discussing alternative ways of writing this.

Reply:
Thanks again for the clarification - yeah, given the reasons you've mentioned above I think this looks good!